library/core/src/slice/specialize.rs

```diff
@@ -0,0 +1,47 @@
+use crate::ptr::write_bytes;
+
+pub(super) trait SpecFill<T> {
+    fn spec_fill(&mut self, value: T);
+}
+
+impl<T: Clone> SpecFill<T> for [T] {
+    default fn spec_fill(&mut self, value: T) {
+        if let Some((last, elems)) = self.split_last_mut() {
+            for el in elems {
+                el.clone_from(&value);
+            }
+
+            *last = value
+        }
+    }
+}
+
+impl<T: Copy> SpecFill<T> for [T] {
+    default fn spec_fill(&mut self, value: T) {
+        for item in self.iter_mut() {
+            *item = value;
+        }
+    }
```

Comment:
Instead of specializing separately on `[u8]` below you can instead do a `if mem::size_of::<T>() == 1` here and call memset for any type that's 1 byte wide.

Reply:
Hm, that's another option. Will do that! Thanks.

Reply:
`value` now has 1-byte size. `ptr::write_bytes` only accepts u8 as second argument.
Should I transmute `value` to u8 ?

Reply:
Even if I want to transmute, the compiler forbid me: <https://rust.godbolt.org/z/a3bsYs>
```rust
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> <source>:38:33
   |
38 |                 let value: u8 = std::mem::transmute(value);
   |                                 ^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `T` (this type does not have a fixed size)
   = note: target type: `u8` (8 bits)
```

Reply:
`transmute_copy` should work

Reply:
Neat! That's really work! But I don't know if that those specific specialization for [u8] matters much with https://github.com/rust-lang/rust/pull/81874#discussion_r571980119

Reply:
The advantage is that it doesn't rely on the optimizer, so it'll work in debug mode too or with less powerful backends.

With the `size_of` approach it only has to be implemented once instead of one specialization for each T.

Reply:
I don't have strong opinion on this. I defer that to reviewer to decide.

Reply:
Hi @bjorn3. the8472 and I were discussing if other codegen backends could handle straight
`for loop` to copy item and make it to memcpy when item is 1-byte size.

Do you think cranelift could be benefited from this optimization ^^?